network/network_test.go

```diff
@@ -632,3 +632,60 @@ func TestDialDeletesNonValidators(t *testing.T) {
 	}
 	wg.Wait()
 }
+
+// Test that cancelling the context passed into dial
+// causes dial to return immediately.
+func TestDialContext(t *testing.T) {
+	_, networks, wg := newFullyConnectedTestNetwork(t, []router.InboundHandler{nil})
+	defer wg.Done()
+	network := networks[0]
+
+	dialer := newTestDialer()
+	network.dialer = dialer
+	dynamicNeverDialedIP, neverDialedListener := dialer.NewListener()
+	neverDialedIP := &trackedIP{
+		ip: dynamicNeverDialedIP.IPPort(),
+	}
+	neverDialedNodeID := ids.GenerateTestNodeID()
+
+	// Asset that when [n.onCloseCtx] is cancelled, dial returns immediately.
+	// That is, [neverDialedListener] doesn't accept a connection.
+	network.manuallyTrackedIDs.Add(neverDialedNodeID)
+	network.onCloseCtxCancel()
+	network.dial(ids.EmptyNodeID, neverDialedIP)
+
+	gotNeverDialedIPConn := make(chan struct{})
+	go func() {
+		_, _ = neverDialedListener.Accept()
+		close(gotNeverDialedIPConn)
+	}()
+
+	select {
+	case <-gotNeverDialedIPConn:
+		require.FailNow(t, "unexpectedly connected to peer")
+	default:
+	}
+
+	// Sanity check that when a non-cancelled context is given,
+	// we actually dial the peer.
+	dialedNodeID := ids.GenerateTestNodeID()
+	network.manuallyTrackedIDs.Add(dialedNodeID)
+	dynamicDialedIP, dialedListener := dialer.NewListener()
+	dialedIP := &trackedIP{
+		ip: dynamicDialedIP.IPPort(),
+	}
+
+	network.dial(dialedNodeID, dialedIP)
+
+	gotDialedIPConn := make(chan struct{})
+	go func() {
+		_, _ = dialedListener.Accept()
+		close(gotDialedIPConn)
+	}()
+
+	select {
+	case <-gotNeverDialedIPConn:
+		require.FailNow(t, "unexpectedly connected to peer")
+	case <-gotDialedIPConn:
+	}
```

Comment:
I think this needs to be done first - as the network context is still cancelled here